compiler/rustc_trait_selection/src/opaque_types.rs

```diff
@@ -1098,21 +1096,45 @@ impl<'a, 'tcx> Instantiator<'a, 'tcx> {
         let infcx = self.infcx;
         let tcx = infcx.tcx;
 
-        debug!("instantiate_opaque_types: Opaque(def_id={:?}, substs={:?})", def_id, substs);
-
         // Use the same type variable if the exact same opaque type appears more
-        // than once in the return type (e.g., if it's passed to a type alias).
-        if let Some(opaque_defn) = self.opaque_types.get(&def_id) {
-            debug!("instantiate_opaque_types: returning concrete ty {:?}", opaque_defn.concrete_ty);
-            return opaque_defn.concrete_ty;
+        // than once in a function (e.g., if it's passed to a type alias).
+        if let Some(opaque_defn) = self.opaque_types.get_mut(&def_id) {
+            debug!(?opaque_defn, "found already known concrete type");
+            if opaque_defn.substs.contains(&substs) {
+                // Already seen this concrete type
+                return opaque_defn.concrete_ty;
+            } else {
+                // Don't emit multiple errors for the same set of substs
+                opaque_defn.substs.push(substs);
+                /*opaque_defn.concrete_ty = tcx.ty_error_with_message(
+                    self.value_span,
+                    "defining use generics differ from previous defining use",
+                );*/
```

Comment:
We have a decision to make here. If I use the error type here, then we won't get the diagnostic about any bounds not getting satisfied but just get the "defining use generics differ from previous defining use". cc @rust-lang/wg-diagnostics I think the additional diagnostics do help a little bit (even if their span is horrible right now, but I think I know how to solve that), but they could also be seen as obfuscating the actual problem.

Reply:
I'd say let's leave it as is for now because I don't think I'd be able to understand what's happening otherwise 😬